lucene/core/src/test/org/apache/lucene/index/TestIndexWriter.java

```
@@ -3941,4 +3943,108 @@ public void testRandomOperationsWithSoftDeletes() throws Exception {
+        }
+
+        @Override
+        public Iterable<? extends IndexableField> next() {
+          try {
+            return docs.nextDoc();
+          } catch (IOException e) {
+            throw new UncheckedIOException(e);
+          }
+        }
+      });
+    } catch (IllegalArgumentException e) {
+      assertEquals("RAM used by a single DocumentsWriterPerThread can't exceed: 1MB", e.getMessage());
+      assertNull(w.getTragicException());
+    }
+    w.addDocument(new Document());
+    numDocs++;
+    try (LineFileDocs docs = new LineFileDocs(new Random(seed))) {
+      final int currentNumDocs = numDocs;
+      for (int i = 0; i < currentNumDocs; i++) {
+        w.addDocument(docs.nextDoc());
+        numDocs++;
+      }
+    }
+    w.addDocument(new Document());
+    numDocs++;
+    w.forceMergeDeletes(true);
+    w.commit();
+
+    try (IndexReader reader = DirectoryReader.open(dir)) {
+      assertEquals(numDocs, reader.numDocs());
+    }
+
+
+    w.close();
+    dir.close();
+  }
+
+  @Nightly
+  public void testAddDocumentsMassive() throws Exception {
+    Directory dir = newFSDirectory(createTempDir("addDocumentsMassive"));
+    IndexWriter w = new IndexWriter(dir, new IndexWriterConfig());
+    StringBuilder sb = new StringBuilder();
+    // same doc with lots of duplicated terms to put pressure on FreqProxTermsWriter instead of terms dict.
+    for (int i = 0; i < 1024; i++) {
+      sb.append(i);
+      sb.append(' ');
+    }
+    final AtomicInteger count = new AtomicInteger(0);
+    Document doc = new Document();
+    try {
+      doc.add(new TextField("text", sb.toString(), Field.Store.NO));
+      w.addDocument(doc);
+      w.addDocuments((Iterable<Document>) () -> new Iterator<>() {
+
+        @Override
+        public boolean hasNext() {
+          return count.get() < 100_000_000;
+        }
+
+        @Override
+        public Document next() {
+          if (!hasNext()) {
+            throw new AssertionError();
```

I think we should throw an error in `hasNext` instead?